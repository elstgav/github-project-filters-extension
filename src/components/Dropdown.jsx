import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import App from 'src/lib/App'


export default class Dropdown extends React.Component {
  static propTypes = {
    buttonText:        PropTypes.node.isRequired,
    buttonProps:       PropTypes.object, // eslint-disable-line react/forbid-prop-types
    children:          PropTypes.node.isRequired,
    className:         PropTypes.string,
    dropdownDirection: PropTypes.string,
  }

  static defaultProps = {
    buttonProps:       {},
    className:         null,
    dropdownDirection: 'sw',
  }

  constructor(props) {
    super(props)
    this.state = { isDropdownOpen: false }
  }

  componentDidUpdate() {
    if (this.state.isDropdownOpen) {
      document.addEventListener('click', this.onDocumentClick, { once: true })
    }
  }

  onDropdownClick = () => {
    this.setState(prevState => ({ isDropdownOpen: !prevState.isDropdownOpen }))
  }

  onDocumentClick = (event) => {
    const isDropdownItem = event.target.classList.contains('dropdown-item')
    const isOutsideClick = !this.ref.contains(event.target)

    if (isDropdownItem || isOutsideClick) this.closeDropdown()
  }

  closeDropdown() {
    this.setState({ isDropdownOpen: false })
  }

  saveRef = (ref) => { this.ref = ref }

  render() {
    return (
      <details
        className={classNames(
          `${App.namespace}-dropdown`,
          'details-reset details-overlay',
          'dropdown',
          this.props.className,
        )}
        ref={this.saveRef}
      >
        <summary
          aria-expanded={this.state.isDropdownOpen}
          aria-haspopup="menu"
          className="btn"
          onClick={this.onDropdownClick}
          role="button"
          {...this.props.buttonProps}
        >
          {this.props.buttonText}
        </summary>

        <details-menu className="dropdown-menu-content f5" role="menu" aria-expanded={this.state.isDropdownOpen}>
          <ul className={`dropdown-menu dropdown-menu-${this.props.dropdownDirection}`}>
            {this.props.children}
          </ul>
        </details-menu>
      </details>
    )
  }
}
