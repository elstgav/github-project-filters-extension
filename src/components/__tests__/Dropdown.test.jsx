import React from 'react'
import simulant from 'simulant'

import Dropdown from 'components/Dropdown'
import EnzymeTestWrapper from 'test/support/EnzymeTestWrapper'

const test = new EnzymeTestWrapper(Dropdown)

describe('Dropdown', () => {
  beforeEach(() => {
    test.props = {
      buttonText: 'Open me',
      children:   'Pick me',
    }
  })

  it('displays a <Dropdown> element', () => {
    expect(test.Dropdown.find('.dropdown')).toHaveLength(1)
    expect(test.Dropdown.props().className).toBe('projecthub-dropdown details-reset details-overlay dropdown')
    expect(test.Dropdown.find('.dropdown-menu-content').text()).toBe('Pick me')
    expect(test.Dropdown).toMatchSnapshot()
  })

  it('has a configurable className', () => {
    test.props.className = 'foo'
    expect(test.Dropdown.props().className).toContain('foo')
  })

  it('has a configurable direction', () => {
    test.props.dropdownDirection = 'ne'
    expect(test.Dropdown.find('.dropdown-menu').props().className).toContain('dropdown-menu-ne')
  })

  describe('button', () => {
    it('has configurable content', () => {
      expect(test.Dropdown.find('summary').text()).toBe('Open me')
    })

    it('has a configurable className', () => {
      test.props.buttonProps = { className: 'foobar', id: 'fooButton' }
      expect(test.Dropdown.find('summary').props().className).toBe('foobar')
      expect(test.Dropdown.find('summary').props().id).toBe('fooButton')
    })
  })

  describe('when open', () => {
    beforeEach(() => {
      test.renderMethod = EnzymeTestWrapper.renderMethods.fullDOM
      test.renderOptions = {
        attachTo: document.body.appendChild(document.createElement('div')),
      }

      test.props = {
        buttonText: 'Open me',
        children: (
          <>
            <li><span className="test-content">Pick me</span></li>
            <li><button type="button" className="dropdown-item">Pick me</button></li>
          </>
        ),
      }

      test.Dropdown.setState({ isDropdownOpen: true })
    })

    it('stays open when you click inside the dropdown', () => {
      expect(test.Dropdown.state('isDropdownOpen')).toBe(true)
      simulant.fire(document.body.querySelector('.test-content'), 'click')
      expect(test.Dropdown.state('isDropdownOpen')).toBe(true)
    })

    it('closes itself when you click a dropdown item', () => {
      expect(test.Dropdown.state('isDropdownOpen')).toBe(true)
      simulant.fire(document.querySelector('.dropdown-item'), 'click')
      expect(test.Dropdown.state('isDropdownOpen')).toBe(false)
    })

    it('closes itself when you click outside the dropdown', () => {
      expect(test.Dropdown.state('isDropdownOpen')).toBe(true)
      simulant.fire(document.body, 'click')
      expect(test.Dropdown.state('isDropdownOpen')).toBe(false)
    })
  })

  describe('when clicked', () => {
    it('toggles its open state', () => {
      expect(test.Dropdown.state('isDropdownOpen')).toBe(false)
      test.Dropdown.find('summary').simulate('click')
      expect(test.Dropdown.state('isDropdownOpen')).toBe(true)
      test.Dropdown.find('summary').simulate('click')
      expect(test.Dropdown.state('isDropdownOpen')).toBe(false)
    })
  })
})
