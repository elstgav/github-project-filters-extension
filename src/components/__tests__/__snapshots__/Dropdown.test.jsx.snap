// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown displays a <Dropdown> element 1`] = `
<details
  className="projecthub-dropdown details-reset details-overlay dropdown"
>
  <summary
    aria-expanded={false}
    aria-haspopup="menu"
    className="btn"
    onClick={[Function]}
    role="button"
  >
    Open me
  </summary>
  <details-menu
    className="dropdown-menu-content f5"
    role="menu"
  >
    <ul
      className="dropdown-menu dropdown-menu-sw"
    >
      Pick me
    </ul>
  </details-menu>
</details>
`;
